src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs

```diff
@@ -553,1618 +450,1915 @@ private bool MethodCallOrArrayIndex(AcceptedCharactersInternal acceptedCharacter
+                    return true;
                 }
 
-                if (At(SyntaxKind.Whitespace) && NextIs(SyntaxKind.QuestionMark))
+                if (type != SyntaxKind.Whitespace &&
+                    type != SyntaxKind.NewLine)
                 {
-                    // Only accept the whitespace if we are going to consume the next token.
-                    AcceptAndMoveNext();
+                    expectingDot = false;
+                    currentIdentifierLength += token.Content.Length;
                 }
 
-                Optional(SyntaxKind.QuestionMark); // Nullable
+                return false;
+            });
 
-                if (At(SyntaxKind.Whitespace) && NextIs(SyntaxKind.LeftBracket))
+            identifierLength = currentIdentifierLength;
+            var validQualifiedIdentifier = expectingDot;
+            if (validQualifiedIdentifier)
+            {
+                foreach (var token in tokens)
                 {
-                    // Only accept the whitespace if we are going to consume the next token.
-                    AcceptAndMoveNext();
+                    identifierLength += token.Content.Length;
+                    Accept(token);
                 }
 
-                while (At(SyntaxKind.LeftBracket))
-                {
-                    Balance(BalancingModes.None);
-                    Optional(SyntaxKind.RightBracket);
-                }
                 return true;
             }
             else
             {
-                return false;
-            }
-        }
-
-        private void TypeArgumentList()
-        {
-            Assert(SyntaxKind.LessThan);
-            Balance(BalancingModes.None);
-            Optional(SyntaxKind.GreaterThan);
-        }
-
-        private void UsingStatement(Block block)
-        {
-            Assert(SyntaxKind.LeftParenthesis);
-
-            // Parse condition
-            if (AcceptCondition())
-            {
-                AcceptWhile(IsSpacingToken(includeNewLines: true, includeComments: true));
-
-                // Parse code block
-                ExpectCodeBlock(block);
-            }
-        }
+                PutCurrentBack();
 
-        private void TryStatement(bool topLevel)
-        {
-            Assert(CSharpKeyword.Try);
-            UnconditionalBlock();
-            AfterTryClause();
-            if (topLevel)
-            {
-                CompleteBlock();
-            }
-        }
+                foreach (var token in tokens)
+                {
+                    identifierLength += token.Content.Length;
+                    PutBack(token);
+                }
 
-        private void IfStatement(bool topLevel)
-        {
-            Assert(CSharpKeyword.If);
-            ConditionalBlock(topLevel: false);
-            AfterIfClause();
-            if (topLevel)
-            {
-                CompleteBlock();
+                EnsureCurrent();
+                return false;
             }
         }
 
-        p
```

Comment:
Small enough method, I'd just inline this. The bits in this class are dope, this as its own method stick out as a sore thumb.